analyzers/diagnostics/diagnostics.go

```diff
@@ -13,42 +15,120 @@ import (
 // fieldRegexp is a regular expression for checking the current line is a field or not.
 var fieldRegexp = `\s*(?P<field>.+)\s*=\s*`
 
-// Diagnostic represents a diagnostics reported by the DeepSource CLI validators.
+// FileDiagnostic represents the diagnostics present in a file.
+type FileDiagnostic struct {
+	Filename    string
+	Diagnostics []Diagnostic
+}
+
+// Diagnostic represents a diagnostic reported by the validators.
 type Diagnostic struct {
+	Filename     string
 	Line         int
 	Codeframe    string
 	ErrorMessage string
 }
 
 // GetDiagnostics returns diagnostics as strings.
 func GetDiagnostics(failure validator.ValidationFailure) ([]string, error) {
-	diagnostics := []string{}
+	// Get diagnostics using the file's content.
+	fileDiagnostics := getDiagnosticsFromFile(failure.File, failure.Errors)
 
-	fileContent, err := readFileContent(failure.File)
-	if err != nil {
-		return nil, err
+	// Group diagnostics based on the filename.
+	groupedDiagnostics := groupDiagnostics(fileDiagnostics)
+
+	// Retrieve string representation of diagnostics.
+	diagnostics := constructDiagnostics(groupedDiagnostics)
+
+	return diagnostics, nil
+}
+
+// groupDiagnostics groups diagnostics based on the filename.
+func groupDiagnostics(diagnostics []Diagnostic) []FileDiagnostic {
+	var groupedDiagnostics []FileDiagnostic
+
+	// Using a map is a clean way to group items.
+	// The diagnostics are grouped using the map, with the key being the filename.
```

Comment:
💯 
Nit: Instead of storing the diagnostics data in `[]Diagnostic`, we could have used a map in the `getDiagnosticsFromFile` function itself in order to store the diagnostic data with respect to files. It would be like `map[string][]Diagnostic` as we have made in this function. That could have been returned from that function. What do you think?

Reply:
`getDiagnosticsFromFile` is intended for retrieving diagnostics from a single file. Making it return a map changes the entire retrieval logic at the caller's end, as the caller now needs to know the key (filename) to access the diagnostics.

Reply:
Understood. Looks good then 👍